Fix #73 - enhance doc (#103)
nicovillamonte/pelelajs: docs/core/BINDING_AND_REACTIVITY.md

# Binding & Reactivity Mechanism

PelelaJS uses a fine-grained reactivity system based on JavaScript `Proxy` objects. The goal is to minimize re-renders by updating only the exact DOM nodes that depend on a changed piece of state.

## The Reactivity Core

At the heart of the system are two main actors:

1. **`ReactiveProxy`**: Wraps the ViewModel state. It intercepts `get` and `set` operations.

2. **`DependencyTracker`**: Keeps a mapping between state properties (keys) and the DOM update functions (subscribers) that rely on them.

```mermaid
graph TD
    A[View Model State] -->|Wrapped in| B(Reactive Proxy)
    B -->|Intercepts SET| C{Dependency Tracker}
    C -->|Triggers| D[DOM Update Function 1]
    C -->|Triggers| E[DOM Update Function 2]
```

> When a template is parsed, any interpolation or binding reads from the `ReactiveProxy`. The `get` trap is triggered, allowing the `DependencyTracker` to register the current DOM update function as a subscriber to that specific property.

Reviewer: ```suggestion
When a template is parsed, any binding reads from the `ReactiveProxy`. The `get` trap is triggered, allowing the `DependencyTracker` to register the current DOM update function as a subscriber to that specific property.
```

## Binding Types

The framework supports distinct binding mechanisms to connect the View to the ViewModel:

### 1. One-Way Binding (Model -> View)

> Updates flow exclusively from the model to the view. Used for interpolations (`{{ value }}`) or property bindings (e.g., `value="user.name"` without two-way modifiers).

Reviewer: ```suggestion
Updates flow exclusively from the model to the view. Used by directives like `bind-content` (text), `bind-class` (CSS classes) and `bind-style` (inline styles).
```

No soportamos interpolación con mustache ([`commons/locales/en.ts:51`](https://github.com/uqbar-project/pelelajs/blob/main/packages/core/src/commons/locales/en.ts#L51)).

- **Mechanism:** The DOM node is subscribed to the property's changes. When the property updates, the DOM node's `textContent` or attribute is overwritten.

### 2. Two-Way Binding (Model <-> View)

> Typically used in input elements (`bind:value="user.name"`).

Reviewer: ```suggestion
Typically used in input elements (`bind-value="user.name"`).
```

- **Mechanism:** It sets up a One-Way binding (Model -> View) for the visual representation, and simultaneously attaches a native DOM event listener (like `input` or `change`) to the element. When the event fires, the framework updates the model, completing the cycle.

### 3. Event Binding (View -> Model)

> Used to trigger ViewModel methods from user actions (`click="submit()"`).

Reviewer: ```suggestion
Used to trigger ViewModel methods from user actions (`click="submit"`).
```
